### PR TITLE
feat: hint master

### PR DESCRIPTION
**Pull request by sanxun0325**
**What this PR does**:
use SQL hint to force the sql route to master node
**Which issue(s) this PR fixes**:

Fixes https://github.com/arana-db/arana/issues/263

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by dongzl** [comment]
@sanxun0325  please format the code import, you can use this tools:

```
go get -u github.com/dubbogo/tools/cmd/imports-formatter

imports-formatter
```

**Review by AlexStocks — APPROVED** [review]

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 	"encoding/json"
 	stdErrors "errors"
 	"fmt"
+	"github.com/arana-db/arana/pkg/proto/hint"
```
Comment (jjeffcaii) [inline pkg/runtime/runtime.go:25]: reformat imports.

File: pkg/runtime/namespace/namespace.go [hunk 2]
```diff
@@ -164,6 +164,49 @@ func (ns *Namespace) DB(ctx context.Context, group string) proto.DB {
...
+		target     = 0
+		wrList     = make([]int, 0, len(exist))
+		readDBList = make([]proto.DB, 0, len(exist))
+	)
+	// slave weight w==0 && r>0
+	for _, db := range exist {
+		if db.Weight().W == 0 && db.Weight().R > 0 {
+			wrList = append(wrList, int(db.Weight().R))
+			readDBList = append(readDBList, db)
+		}
```
Comment (jjeffcaii) [inline pkg/runtime/namespace/namespace.go:202]: My mistake, please change the slave logic to `W==0 && R==0`.

**Review by jjeffcaii — CHANGES_REQUESTED** [review]
PS: do not change existing integration tests, add a new ut.

File: pkg/runtime/namespace/namespace.go (hunk 2 above)
Comment (dongzl) [reply pkg/runtime/namespace/namespace.go:202]: > My mistake, please change the slave logic to `W==0 && R==0`.

why the slave node  `R==0`?

Comment (jjeffcaii) [reply pkg/runtime/namespace/namespace.go:202]: The final conclusion:
- MASTER: w > 0 && r >= 0
- SLAVE: w == 0 && r >= 0


PS: the `r==0` has high priority.

File: pkg/runtime/runtime.go (hunk 1 above)
Comment (sanxun0325) [reply pkg/runtime/runtime.go]: done

File: pkg/runtime/namespace/namespace.go (hunk 2 above)
Comment (sanxun0325) [reply pkg/runtime/namespace/namespace.go]: done

File: test/integration_test.go [hunk 3]
```diff
@@ -69,7 +69,7 @@ func (s *IntegrationSuite) TestBasicTx() {
 	assert.NoError(t, err)
 	assert.Equal(t, int64(1), affected)
 
-	row := tx.QueryRow("SELECT COUNT(1) FROM sequence WHERE name=?", name)
+	row := tx.QueryRow("/*A! master */ SELECT COUNT(1) FROM sequence WHERE name=?", name)
```
Comment (jjeffcaii) [inline test/integration_test.go:72]: Remove hint after change weight to `r0w0`, same as below.

File: conf/config.yaml [hunk 4]
```diff
@@ -52,6 +52,15 @@ data:
               weight: r10w10
               parameters:
                 maxAllowedPacket: 256M
+            - name: node0_r_0
+              host: arana-mysql
+              port: 3306
+              username: root
+              password: "123456"
+              database: employees_0000_r
+              weight: r10w0
```
[the remark this answers is not in the document]
Comment (sanxun0325) [reply conf/config.yaml]: done

**Review by dongzl — APPROVED** [review]

**Review by jjeffcaii — APPROVED** [review]

**Comment by jjeffcaii** [comment]
Please fix ut failure.

**Review by dongzl — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Comment by jjeffcaii** [comment]
Fix #263

**Pull request merged**